### PR TITLE
i#6548: drmemtrace does not create the required pipe file if not present.

### PR DESCRIPTION
**Pull request by philramsey-arm**
Update os_open() to specify the O_CREAT flag when opening a file with the OS_OPEN_WRITE_ONLY flag.

Also refactor this function to make it more maintainable.

Output a usage warning if drmemtrace has to create the pipe file

Issue: #6548

### CODE REVIEW

File: clients/drcachesim/tracer/tracer.cpp [hunk 1]
```diff
@@ -2336,7 +2336,14 @@ drmemtrace_client_main(client_id_t id, int argc, const char *argv[])
             DR_ASSERT(false);
 #ifdef UNIX
         /* we want an isolated fd so we don't use ipc_pipe.open_for_write() */
-        int fd = dr_open_file(ipc_pipe.get_pipe_path().c_str(), DR_FILE_WRITE_ONLY);
+        const char *pipe_path = ipc_pipe.get_pipe_path().c_str();
+        if (!dr_file_exists(pipe_path))
```
Comment (derekbruening) [inline clients/drcachesim/tracer/tracer.cpp:2340]: style: {} around multi-line body (vera++ checks should complain about this)

File: clients/drcachesim/tracer/tracer.cpp [hunk 2]
```diff
@@ -2336,7 +2336,14 @@ drmemtrace_client_main(client_id_t id, int argc, const char *argv[])
...
+            NOTIFY( 0,
+                    "drmemtrace: creating pipe file at \"%s\". Use \"-offline\" mode if "
```
Comment (derekbruening) [inline clients/drcachesim/tracer/tracer.cpp:2342]: Should it say "attempting to open write end of pipe for online analysis but pipe does not exist"?

File: clients/drcachesim/tracer/tracer.cpp [hunk 3]
```diff
@@ -2336,7 +2336,14 @@ drmemtrace_client_main(client_id_t id, int argc, const char *argv[])
             DR_ASSERT(false);
```
Comment (derekbruening) [inline clients/drcachesim/tracer/tracer.cpp:2336]: > i#6548 drmemtrace does not create the required pipe file if not present.

nit: We put a colon after the issue number in the title: `s/i#6548 d/i#6548: d/`

**Review by derekbruening — APPROVED** [review]

File: clients/drcachesim/tracer/tracer.cpp (hunk 2 above)
Comment (derekbruening) [reply clients/drcachesim/tracer/tracer.cpp:2342]: Maybe with a `WARNING: ` prefix too?

**Comment by AssadHashmi** [comment]
x86 regressions related to https://github.com/DynamoRIO/dynamorio/issues/6417.

**Pull request merged**